File: C++/7-Trees/Algorithms/2-Breadth-FirstSearch(BFS)/Breadth-FirstSearch(BFS)Pro.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>
#include <limits> // Required for numeric_limits

// Basic structure for a tree node
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

// Function to build a tree from a vector of values
// nullptr values in the vector represent missing nodes
TreeNode* buildTree(const std::vector<int>& values) {
    if (values.empty()) return nullptr;

    TreeNode* root = new TreeNode(values[0]);
    std::queue<TreeNode*> q;
    q.push(root);
    int i = 1;

    while (!q.empty() && i < values.size()) {
        TreeNode* current = q.front();
        q.pop();

        // Left child
        if (values[i] != std::numeric_limits<int>::max()) { // Use max int to represent null
            current->left = new TreeNode(values[i]);
            q.push(current->left);
        }
        i++;

        if (i < values.size() && values[i] != std::numeric_limits<int>::max()) {
            current->right = new TreeNode(values[i]);
            q.push(current->right);
        }
        i++;
    }
    return root;
}

// Function to print the tree in level order (BFS) - for verifying the tree structure
void printLevelOrder(TreeNode* root) {
    if (!root) {
        std::cout << "Tree is empty." << std::endl;
        return;
    }

    std::queue<TreeNode*> q;
    q.push(root);
    std::cout << "Level Order Traversal: ";
    while (!q.empty()) {
        TreeNode* current = q.front();
        q.pop();
        std::cout << current->val << " ";
        if (current->left) q.push(current->left);
        if (current->right) q.push(current->right);
    }
    std::cout << std::endl;
}

// 1. Basic Breadth-First Search (BFS)
//   - Concept: Simplest form of BFS, visiting nodes level by level.
//   - Real-world application: Finding the shortest path in an unweighted graph,
//     network routing (finding the closest servers), web crawler to explore pages.
void basicBFS(TreeNode* root) {
    if (!root) return;

    std::queue<TreeNode*> q;
    q.push(root);

    std::cout << "Basic BFS: ";
    while (!q.empty()) {
        TreeNode* current = q.front();
        q.pop();
        std::cout << current->val << " ";
        if (current->left) q.push(current->left);
        if (current->right) q.push(current->right);
    }
    std::cout << std::endl;
}

// 2. BFS with Level Separation
//   - Concept:  Processes the tree level by level, keeping track of node counts for each level.
//   - Real-world application:  Calculating the height of a tree,
//      processing nodes in a specific order for UI rendering (e.g., layers),
//      analyzing tree structure (number of nodes per level).
void bfsWithLevelSeparation(TreeNode* root) {
    if (!root) return;

    std::queue<TreeNode*> q;
    q.push(root);
    int level = 0;

    std::cout << "BFS with Level Separation:\n";
    while (!q.empty()) {
        int levelSize = q.size();
        std::cout << "Level " << level++ << ": ";
        for (int i = 0; i < levelSize; ++i) {
            TreeNode* current = q.front();
            q.pop();
            std::cout << current->val << " ";
            if (current->left) q.push(current->left);
            if (current->right) q.push(current->right);
        }
        std::cout << std::endl;
    }
}

// 3. BFS for Finding a Specific Node
//    - Concept:  Modifies BFS to search for a target value within the tree.
//    - Real-world application:  Searching for a specific file or directory in a file system,
//       locating a particular user in a hierarchical organization structure,
//       finding a specific product in a category tree.
bool bfsFindNode(TreeNode* root, int target) {
    if (!root) return false;

    std::queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* current = q.front();
        q.pop();
        if (current->val == target) {
            std::cout << "Found node with value " << target << " using BFS.\n";
            return true;
        }
        if (current->left) q.push(current->left);
        if (current->right) q.push(current->right);
    }
    std::cout << "Node with value " << target << " not found using BFS.\n";
    return false;
}

// 4. BFS to find the Shortest Path (in an unweighted tree, path length is # of edges)
//    - Concept: Uses BFS to find the shortest path from the root to a target node.
//    - Real-world application: Network routing (finding the route with the fewest hops),
//      game AI (finding the shortest path to an objective),
//      social network analysis (finding the shortest connection between two people).
int bfsShortestPath(TreeNode* root, int target) {
    if (!root) return -1;

    std::queue<std::pair<TreeNode*, int>> q; // Store node and its distance from root
    q.push({root, 0});
    std::vector<int> visited;

    while (!q.empty()) {
        TreeNode* current = q.front().first;
        int distance = q.front().second;
        q.pop();

        // Check if the node has been visited
        bool isVisited = false;
        for (int v : visited) {
            if (v == current->val) {
                isVisited = true;
                break;
            }
        }
        if (isVisited) continue;
        visited.push_back(current->val);

        if (current->val == target) {
            std::cout << "Shortest path to node " << target << " is " << distance << ".\n";
            return distance;
        }
        if (current->left) q.push({current->left, distance + 1});
        if (current->right) q.push({current->right, distance + 1});
    }
    std::cout << "Node " << target << " not found.\n";
    return -1;
}

// 5.  Iterative Level Order Traversal with Zigzag Pattern
//     - Concept: Performs a level-order traversal but alternates the direction
//       of visiting nodes between left-to-right and right-to-left.
//     - Real-world application:  Displaying search results in an alternating pattern for better UI readability,
//        processing data in a specific order for optimization,
//        text justification algorithms.
void zigzagLevelOrder(TreeNode* root) {
    if (!root) return;

    std::queue<TreeNode*> q;
    q.push(root);
    bool leftToRight = true; // Direction flag

    std::cout << "Zigzag Level Order Traversal:\n";
    while (!q.empty()) {
        int levelSize = q.size();
        std::vector<int> currentLevel(levelSize); // Store nodes at current level

        for (int i = 0; i < levelSize; ++i) {
            TreeNode* current = q.front();
            q.pop();
            // Store the node's value in the correct order based on direction
            if (leftToRight) {
                currentLevel[i] = current->val;
            } else {
                currentLevel[levelSize - 1 - i] = current->val;
            }

            if (current->left) q.push(current->left);
            if (current->right) q.push(current->right);
        }
        // Print the current level
        for (int val : currentLevel) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
        leftToRight = !leftToRight; // Flip the direction
    }
}

int main() {
    // Example tree construction:
    //       1
    //     /   \
    //    2     3
    //   / \   / \
    //  4   5 6   7
    //
    //  Represented as vector, using max int for null: [1, 2, 3, 4, 5, 6, 7]
    std::vector<int> treeValues = {1, 2, 3, 4, 5, 6, 7};
    TreeNode* root = buildTree(treeValues);
    printLevelOrder(root); // Verify tree structure.

    // Test the BFS functions
    basicBFS(root);
    bfsWithLevelSeparation(root);
    bfsFindNode(root, 5);
    bfsFindNode(root, 8); // Test case where node is not found.
    bfsShortestPath(root, 7);
    bfsShortestPath(root, 8); // Test case where node is not found.
    zigzagLevelOrder(root);

    // Example of a different tree, including nullptr/null nodes
    //       1
    //     /   \
    //    2     3
    //   / \     \
    //  4   5     7
    std::vector<int> treeValues2 = {1, 2, 3, 4, 5, std::numeric_limits<int>::max(), 7};
    TreeNode* root2 = buildTree(treeValues2);
    printLevelOrder(root2);

    basicBFS(root2);
    bfsWithLevelSeparation(root2);
    bfsFindNode(root2, 7);
    bfsFindNode(root2, 6);
    bfsShortestPath(root2, 7);
    bfsShortestPath(root2, 6);
    zigzagLevelOrder(root2);

    // Remember to free the allocated memory for the tree nodes if needed
    // (not shown in this example for brevity, but important in real applications).
    return 0;
}

File: Java/7-Trees/Algorithms/2-Breadth-FirstSearch(BFS)/Breadth-FirstSearch(BFS)Pro.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.HashSet;

public class TreeBFS {

    public static void main(String[] args) {
        // Create a sample binary tree for testing
        TreeNode<Integer> root = new TreeNode<>(1);
        root.left = new TreeNode<>(2);
        root.right = new TreeNode<>(3);
        root.left.left = new TreeNode<>(4);
        root.left.right = new TreeNode<>(5);
        root.right.left = new TreeNode<>(6);
        root.right.right = new TreeNode<>(7);
        root.left.left.left = new TreeNode<>(8);
        root.left.left.right = new TreeNode<>(9);

        System.out.println("Tree structure for testing:");
        printTree(root); // Helper method to visualize the tree structure

        System.out.println("\n1. Basic BFS Traversal:");
        List<Integer> bfsResult = basicBFS(root);
        System.out.println(bfsResult); // Output: [1, 2, 3, 4, 5, 6, 7, 8, 9]

        System.out.println("\n2. BFS with Level Information:");
        List<List<Integer>> bfsWithLevelResult = bfsWithLevel(root);
        System.out.println(bfsWithLevelResult); // Output: [[1], [2, 3], [4, 5, 6, 7], [8, 9]]

        System.out.println("\n3. BFS for Searching a Value (Returns true if found):");
        boolean found = bfsSearch(root, 5);
        System.out.println("Found 5? " + found); // Output: true
        boolean notFound = bfsSearch(root, 10);
        System.out.println("Found 10? " + notFound); // Output: false

        System.out.println("\n4. BFS to find maximum width of tree:");
        int maxWidth = findMaxWidth(root);
        System.out.println("Maximum width of tree is: " + maxWidth);

        System.out.println("\n5. BFS for Tree Serialization/Deserialization (Conceptual - prints level order):");
        //  This shows the level order traversal, which is the core of serialization/deserialization
        List<Integer> serialized = serializeTree(root);
        System.out.println("Serialized (Level Order): " + serialized);
        // Deserialization would involve reconstructing the tree from this level order,
        // but that's a more complex process and is beyond the scope of a simple BFS example.
        // It's included as a conceptual application.
    }

    // Helper method to print the tree structure (for visualization)
    public static void printTree(TreeNode<?> root) {
        if (root == null) return;
        Queue<TreeNode<?>> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            for (int i = 0; i < levelSize; i++) {
                TreeNode<?> node = queue.poll();
                System.out.print(node.val + " ");
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            System.out.println(); // Move to the next level
        }
    }

    /**
     * 1. Basic Breadth-First Search (BFS) Traversal
     *
     * -   Performs a standard BFS traversal of the binary tree.
     * -   Visits nodes level by level, from left to right.
     * -   Uses a queue to maintain the order of nodes to visit.
     * -   Returns a list containing the values of the nodes in the order they were visited.
     * -   Real-world application:
     * -   Crawling a website:  BFS can be used to traverse the pages of a website,
     * starting from the home page and following links to other pages.
     *
     * @param root The root of the binary tree to traverse.
     * @return A list of node values in BFS order.  Returns an empty list if the root is null.
     */
    public static List<Integer> basicBFS(TreeNode<Integer> root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<TreeNode<Integer>> queue = new LinkedList<>();
        queue.offer(root); // Add the root node to the queue

        while (!queue.isEmpty()) {
            TreeNode<Integer> node = queue.poll(); // Dequeue the front node
            result.add(node.val); // Process the node's value

            // Enqueue the left and right children, if they exist
            if (node.left != null) {
                queue.offer(node.left);
            }
            if (node.right != null) {
                queue.offer(node.right);
            }
        }
        return result;
    }

    /**
     * 2. BFS Traversal with Level Information
     *
     * -   Performs a BFS traversal, but also keeps track of the level of each node.
     * -   Returns a list of lists, where each inner list represents a level of the tree.
     * -   Nodes within each level are ordered from left to right.
     * -  Real-world application:
     * - Network Routing:  Finding the shortest path between two nodes in a network
     * can be implemented using a modified BFS that considers network topology.
     *
     * @param root The root of the binary tree.
     * @return A list of lists, where each inner list contains the node values at that level.
     * Returns an empty list if the root is null.
     */
    public static List<List<Integer>> bfsWithLevel(TreeNode<Integer> root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<TreeNode<Integer>> queue = new LinkedList<>();
        queue.offer(root);
        int level = 0;

        while (!queue.isEmpty()) {
            int levelSize = queue.size(); // Number of nodes at the current level
            List<Integer> currentLevel = new ArrayList<>();
            for (int i = 0; i < levelSize; i++) {
                TreeNode<Integer> node = queue.poll();
                currentLevel.add(node.val);

                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
            result.add(currentLevel);
            level++;
        }
        return result;
    }

    /**
     * 3. BFS for Searching a Value
     *
     * -   Performs a BFS traversal to search for a specific value within the tree.
     * -   Returns true if the value is found, false otherwise.
     * -   This is an adaptation of BFS for a search problem, not just traversal.
     * -   Real-world application:
     * -   Social Network Search: Searching for a user in a social network, where
     * nodes represent users and edges represent connections.
     *
     * @param root The root of the tree to search in.
     * @param target The value to search for.
     * @return True if the value is found, false otherwise.
     */
    public static boolean bfsSearch(TreeNode<Integer> root, int target) {
        if (root == null) {
            return false;
        }

        Queue<TreeNode<Integer>> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode<Integer> node = queue.poll();
            if (node.val == target) {
                return true; // Value found!
            }

            if (node.left != null) {
                queue.offer(node.left);
            }
            if (node.right != null) {
                queue.offer(node.right);
            }
        }
        return false; // Value not found after traversing the entire tree
    }

    /**
     * 4. BFS to find maximum width of tree
     *
     * - Uses BFS to find the maximum width of the binary tree.
     * - The width of a level is the number of nodes at that level.
     * - Real world application:
     * - UI layout: Determining the maximum width of a widget or element
     * in a hierarchical user interface structure.
     *
     * @param root - root node of the tree
     * @return - returns the maximum width
     */

    public static int findMaxWidth(TreeNode<Integer> root) {
        if (root == null) {
            return 0;
        }

        Queue<TreeNode<Integer>> queue = new LinkedList<>();
        queue.offer(root);
        int maxWidth = 0;

        while (!queue.isEmpty()) {
            int levelSize = queue.size(); // Get the size of the current level
            maxWidth = Math.max(maxWidth, levelSize); // Update maxWidth if current level is wider

            for (int i = 0; i < levelSize; i++) {
                TreeNode<Integer> node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
        }
        return maxWidth;
    }

    /**
     * 5. BFS for Tree Serialization/Deserialization (Conceptual)
     *
     * -   Demonstrates how BFS can be used conceptually for tree serialization.
     * -   Serialization is the process of converting a tree into a linear format
     * (e.g., a string or a list) so that it can be stored or transmitted.
     * -   BFS provides a level-order traversal, which is a common way to serialize a tree.
     * -   Deserialization is the reverse process: constructing the tree from the linear format.
     * -   This method focuses on the serialization aspect (using BFS to get level order).
     * -   Real-world application:
     * -   Data storage and retrieval:  Serializing a tree allows it to be stored in a file
     * or database, and then deserialized to reconstruct the tree in memory when needed.
     * -   Distributed systems:  Trees can be serialized to be transmitted over a network
     * between different systems.
     *
     * @param root The root of the tree to serialize.
     * @return A list of Integers representing the level order traversal of the tree.
     */
    public static List<Integer> serializeTree(TreeNode<Integer> root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<TreeNode<Integer>> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode<Integer> node = queue.poll();
            result.add(node.val); // Add the value, NOT the node.

            if (node.left != null) {
                queue.offer(node.left);
            }
            if (node.right != null) {
                queue.offer(node.right);
            }
        }
        return result;
    }
}

// Definition of a generic TreeNode class.  This is used across all the examples.
class TreeNode<T> {
    T val;
    TreeNode<T> left;
    TreeNode<T> right;

    public TreeNode(T val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }

    public TreeNode(T val, TreeNode<T> left, TreeNode<T> right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    @Override
    public String toString() {
        return "TreeNode{" +
                "val=" + val +
                ", left=" + (left == null ? "null" : left.val) + // Avoid infinite recursion in toString
                ", right=" + (right == null ? "null" : right.val) +
                "}";
    }
}